File: example/cpp/exampleRemoteControlboard/ExampleRemoteControlboard.cpp
// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-

#include "ExampleRemoteControlboard.hpp"

namespace roboticslab
{

int ExampleRemoteControlboard::run(int argc, char **argv)
{
    //-- Init
    yarp::os::ResourceFinder rf;
    rf.setVerbose(true);
    rf.setDefaultContext("exampleRemoteControlboard");
    rf.setDefaultConfigFile("exampleRemoteControlboard.ini");
    rf.configure(argc, argv);

    std::string robot = rf.check("robot",yarp::os::Value(DEFAULT_ROBOT),"name of /robot to be used").asString();
    printf("\t--robot: %s [%s]\n",robot.c_str(),DEFAULT_ROBOT);

    printf("Note: requires a running instance of roboticslabSim\n");
    if (!yarp::os::Network::checkNetwork())
    {
        printf("Please start a yarp name server first\n");
        return 1;
    }

    //-- Configure device
    yarp::os::Property options; //create an instance of Property, a nice YARP class for storing name-value (key-value) pairs
    options.put("device","remote_controlboard"); //we add a name-value pair that indicates the YARP device
    options.put("remote",robot); //we add info on to whom we will connect
    options.put("local","/local"); //we add info on how we will call ourselves on the YARP network
    dd.open(options); //Configure the YARP multi-use driver with the given options
    if( ! dd.isValid() )
    {
        printf("%s not available.\n", robot.c_str());
        dd.close();
        yarp::os::Network::fini(); //disconnect from the YARP network
        return 1;
    }

    //-- View interfaces
    if ( ! dd.view(pos) )  // connect 'pos' interface to 'dd' device
    {
        printf("[error] Problems acquiring position interface\n");
        return 1;
    }
    if ( ! dd.view(pos2) )  // connect 'pos2' interface to 'dd' device
    {
        printf("[error] Problems acquiring position interface\n");
        return 1;
    }
    printf("[success] Acquired position interface\n");

    if ( ! dd.view(enc) ) // connect 'enc' interface to 'dd' device
    {
        printf("[error] Problems acquiring encoder interface\n");
        return 1;
    }
    printf("[success] Acquired encoder interface\n");

    if ( ! dd.view(vel) ) // connect 'vel' interface to 'dd' device
    {
        printf("[error] Problems acquiring velocity interface\n");
        return 1;
    }
    printf("[success] Acquired velocity interface\n");

    //-- Start

    /*printf("setPositionMode()\n");
    pos->setPositionMode(); //use the position object to set the device to position mode (as opposed to velocity mode)

    printf("setRefSpeed(0,5)\n");
    pos->setRefSpeed(0,5);

    printf("setRefAcceleration(0,5)\n");
    pos->setRefAcceleration(0,5);

    printf("positionMove(0,-10)\n");
    pos->positionMove(0, -10);

    printf("Wait to reach");
    bool done = false;
    do
    {
        yarp::os::Time::delay(0.1);
        pos->checkMotionDone( & done );
        printf(".");
        fflush(stdout);
    }
    while( ! done );
    printf("\n");

    double d;
    enc->getEncoder(0,&d);
    printf("getEncoder(0) -> is at: %f\n", d);

    printf("setVelocityMode()\n");
    vel->setVelocityMode(); //use the object to set the device to velocity mode (as opposed to position mode)

    printf("velocityMove(0,5)\n");
    vel->velocityMove(0,5);

    printf("Delaying 2 seconds...\n");
    yarp::os::Time::delay(2);

    printf("velocityMove(0,0) <- stop\n");
    vel->velocityMove(0,0);

    printf("setPositionMode()\n");
    pos->setPositionMode(); //use the position object to set the device to position mode (as opposed to velocity mode)*/

    /*{
        printf("positionMove() <- -3,-3\n");
        std::vector<double> q(2,0.0);
        q[0] = -3.0;
        q[1] = -3.0;
        pos->positionMove( q.data() );
        printf("Wait to reach");
        bool done = false;
        while(!done)
        {
            pos->checkMotionDone( & done );
            printf(".");
            fflush(stdout);
            yarp::os::Time::delay(0.1);
        }
        printf("\n");

        std::vector<double> d(2,0.0);
        enc->getEncoders( d.data() );
        printf("getEncoders() -> is at: %f %f\n", d[0], d[1]);
    }

    {
        printf("positionMove() <- -0,-0\n");
        std::vector<double> q(2,0.0);
        q[0] = -3.0;
        q[1] = -3.0;
        pos->positionMove( q.data() );
        printf("Wait to reach");
        bool done = false;
        while(!done)
        {
            pos->checkMotionDone( & done );
            printf(".");
            fflush(stdout);
            yarp::os::Time::delay(0.1);
        }
        printf("\n");

        std::vector<double> d(2,0.0);
        enc->getEncoders( d.data() );
        printf("getEncoders() -> is at: %f %f\n", d[0], d[1]);
    }*/

    {
        printf("positionMove() <- -3,-3\n");
        std::vector<double> q(2,0.0);
        std::vector<int> mask(2,1);
        q[0] = -3.0;
        q[1] = -3.0;
        pos2->positionMove( 2, mask.data(), q.data() );
        printf("Wait to reach");
        bool done = false;
        while(!done)
        {
            pos->checkMotionDone( & done );
            printf(".");
            fflush(stdout);
            yarp::os::Time::delay(0.1);
        }
        printf("\n");

        std::vector<double> d(2,0.0);
        enc->getEncoders( d.data() );
        printf("getEncoders() -> is at: %f %f\n", d[0], d[1]);
    }


    dd.close();

    return 0;
}

} //namespace roboticslab

File: example/python/exampleRemoteControlboard.py
#! /usr/bin/env python

##
# @ingroup yarp_devices_examples_py
# \defgroup exampleRemoteControlboardPy exampleRemoteControlboard.py
#
# This example connects to a running \ref controlboardWrapper to move in Joint space.
#
# <b>Legal</b> 
#
# Copyright: (C) 2017 Universidad Carlos III de Madrid
#
# Author: Juan G Victores
#
# CopyPolicy: Released under the terms of the LGPLv2.1 or later, see license/LGPL.TXT
#
# <b>Installation</b>
#
# We basically just need to give the module running (execution) permissions. In Ubuntu/Linux:
#
#\verbatim
#chmod +x $YARP_DEVICES_ROOT/example/python/exampleRemoteControlboard.py
#\endverbatim
#
# <b>Running</b> (assuming correct installation)
#
#\verbatim
#$YARP_DEVICES_ROOT/example/python/exampleRemoteControlboard.py
#\endverbatim
#
# <b>Modify</b>
#
# This file can be edited at $YARP_DEVICES_ROOT/example/python/exampleRemoteControlboard.py
#

import yarp  # imports YARP
yarp.Network.init()  # connect to YARP network
if yarp.Network.checkNetwork() != True:  # let's see if there was actually a reachable YARP network
    print '[error] Please try running yarp server'  # tell the user to start one with 'yarp server' if there isn't any
    quit()
options = yarp.Property()  # create an instance of Property, a nice YARP class for storing name-value (key-value) pairs
options.put('device','remote_controlboard')  # we add a name-value pair that indicates the YARP device
options.put('remote','/robotName')  # we add info on to whom we will connect
options.put('local','/exampleRemoteControlboard')  # we add info on how we will call ourselves on the YARP network
dd = yarp.PolyDriver(options)  # create a YARP multi-use driver with the given options

pos = dd.viewIPositionControl()  # make a position controller object we call 'pos'
vel = dd.viewIVelocityControl()  # make a velocity controller object we call 'vel'
enc = dd.viewIEncoders()  # make an encoder controller object we call 'enc'
mode = dd.viewIControlMode()  # make a operation mode controller object we call 'mode'

enc = dd.viewIEncoders()  # make an encoder controller object we call 'enc'
axes = enc.getAxes()  # retrieve number of joints

# use the object to set the device to position mode (as opposed to velocity mode)(note: stops the robot)
for j in range(axes):
    mode.setPositionMode(j)

print 'test positionMove(1,-35) -> moves motor 1 (start count at motor 0) to -35 degrees'
pos.positionMove(1,-35)

print 'test delay(5)'
yarp.Time.delay(5)

v = yarp.DVector(axes)  # create a YARP vector of doubles the size of the number of elements read by enc, call it 'v'
enc.getEncoders(v)  # read the encoder values and put them into 'v'
print 'v[1] is: ' + str(v[1])  # print element 1 of 'v', note that motors and encoders start at 0

# use the object to set the device to velocity mode (as opposed to position mode)
for j in range(1,axes):
    mode.setPositionMode(j)

print 'test velocityMove(0,10) -> moves motor 0 (start count at motor 0) at 10 degrees per second'
vel.velocityMove(0,10)

print 'test delay(5)'
yarp.Time.delay(5)

vel.velocityMove(0,0)  # stop the robot
dd.close()

yarp.Network.fini()  # disconnect from the YARP network
